crates/bevy_core_pipeline/src/smaa/smaa.wgsl

```diff
@@ -0,0 +1,1122 @@
+// above, in "SMAA Presets".)
+
+/**
+ * SMAA_CORNER_ROUNDING specifies how much sharp corners will be rounded.
+ *
+ * Range: [0, 100]
+ *
+ * Define SMAA_DISABLE_CORNER_DETECTION to disable corner processing.
+ */
+// (In the WGSL version of this shader, `SMAA_CORNER_ROUNDING` is set above, in
+// "SMAA Presets".)
+
+/**
+ * If there is an neighbor edge that has SMAA_LOCAL_CONTRAST_FACTOR times
+ * bigger contrast than current edge, current edge will be discarded.
+ *
+ * This allows to eliminate spurious crossing edges, and is based on the fact
+ * that, if there is too much contrast in a direction, that will hide
+ * perceptually contrast in the other neighbors.
+ */
+const SMAA_LOCAL_CONTRAST_ADAPTATION_FACTOR: f32 = 2.0;
+
+//-----------------------------------------------------------------------------
+// Non-Configurable Defines
+
+const SMAA_AREATEX_MAX_DISTANCE: f32 = 16.0;
+const SMAA_AREATEX_MAX_DISTANCE_DIAG: f32 = 20.0;
+const SMAA_AREATEX_PIXEL_SIZE: vec2<f32> = (1.0 / vec2<f32>(160.0, 560.0));
+const SMAA_AREATEX_SUBTEX_SIZE: f32 = (1.0 / 7.0);
+const SMAA_SEARCHTEX_SIZE: vec2<f32> = vec2(66.0, 33.0);
+const SMAA_SEARCHTEX_PACKED_SIZE: vec2<f32> = vec2(64.0, 16.0);
+
+#ifndef SMAA_DISABLE_CORNER_DETECTION
+const SMAA_CORNER_ROUNDING_NORM: f32 = f32(SMAA_CORNER_ROUNDING) / 100.0;
+#endif  // SMAA_DISABLE_CORNER_DETECTION
+
+//-----------------------------------------------------------------------------
+// Porting Functions
+
+#ifdef SMAA_BLENDING_WEIGHT_CALCULATION
+fn sample_edges_level_zero_offset(coord: vec2<f32>, offset: vec2<i32>) -> vec4<f32> {
+    return textureSampleLevel(
+        edges_texture,
+        edges_sampler,
+        coord + vec2<f32>(offset) * smaa_info.rt_metrics.xy,
+        0.0
+    );
+}
+#endif  // SMAA_BLENDING_WEIGHT_CALCULATION
+
+//-----------------------------------------------------------------------------
+// WGSL-Specific Functions
+
+// This vertex shader produces the following, when drawn using indices 0..3:
+//
+//  1 |  0-----x.....2
+//  0 |  |  s  |  . ´
+// -1 |  x_____x´
+// -2 |  :  .´
+// -3 |  1´
+//    +---------------
+//      -1  0  1  2  3
+//
+// The axes are clip-space x and y. The region marked s is the visible region.
+// The digits in the corners of the right-angled triangle are the vertex
+// indices.
+//
+// The top-left has UV 0,0, the bottom-left has 0,2, and the top-right has 2,0.
+// This means that the UV gets interpolated to 1,1 at the bottom-right corner
+// of the clip-space rectangle that is at 1,-1 in clip space.
+fn calculate_vertex_varyings(vertex_index: u32) -> VertexVaryings {
+    // See the explanation above for how this works
+    let uv = vec2<f32>(f32(vertex_index >> 1u), f32(vertex_index & 1u)) * 2.0;
+    let clip_position = vec2<f32>(uv * vec2<f32>(2.0, -2.0) + vec2<f32>(-1.0, 1.0));
+
+    return VertexVaryings(clip_position, uv);
+}
+
+fn texture_offset(coords: vec2<f32>, offset: vec2<i32>) -> vec2<f32> {
```

Comment:
This isn't needed because [`textureSampleLevel`](https://gpuweb.github.io/gpuweb/wgsl/#texturesamplelevel) has an offset parameter.